### DISCUSSION TITLE
No access to response after receiving response and then RST_STREAM

### DISCUSSION THREAD

**Discussion by chradcliffe** [encode/httpx#1625] [Potential Issue]
We're using httpx for some end-to-end testing of an HTTP/2 proxy server, and we ran into an issue where httpx throws an exception on a RST_STREAM after having received the response headers and body. The sequence is as follows:
1. httpx sends a POST request with a large payload (larger than the stream window size). The request headers are sent along with part of the payload (up to the window size).
2. The proxy server logic sees the request headers and decides to respond before having seen the entirety of the POST payload.
3. The HTTP/2 stack sends a HEADERS frame and a DATA frame in response. It also sends a RST_STREAM frame with error code 0x0 (NO_ERROR) to stop the client from sending the remaining POST payload.
4. httpx raises a `RemoteProtocolError` exception, but the response headers/data is unavailable, despite having been received by the library.

I turned on trace logging for httpx and this happens just prior to the exception being thrown:
```
TRACE [2021-05-04 19:34:44] httpcore._async.http2 - receive_event stream_id=1 event=<ResponseReceived stream_id:1, headers:[(b':status', b'200'), (b'content-type', b'text/html; charset=UTF-8'), (b'content-length', b'13'), (b'cache-control', b'no-cache'), (b'access-control-allow-origin', b'*'), (b'date', b'Tue, 04 May 2021 19:34:44 GMT')]>
TRACE [2021-05-04 19:34:44] httpcore._async.http2 - receive_event stream_id=1 event=<DataReceived stream_id:1, flow_controlled_length:13, data:[redacted]>
TRACE [2021-05-04 19:34:44] httpcore._async.http2 - receive_event stream_id=1 event=<StreamEnded stream_id:1>
TRACE [2021-05-04 19:34:44] httpcore._async.http2 - receive_event stream_id=1 event=<StreamReset stream_id:1, error_code:ErrorCodes.NO_ERROR, remote_reset:True>
[ backtrace here ]
httpx.RemoteProtocolError: <StreamReset stream_id:1, error_code:ErrorCodes.NO_ERROR, remote_reset:True>
```

Given that the entire POST is not received, I think it's a reasonable approach to throw an exception when the stream is reset in this way; however, I would hope to be able to inspect the response that was received on this sort of error.

By way of comparison, curl will print the response but exit with a non-zero status under these conditions. Firefox will dutifully display the response as well.

Naively (not being overly familiar with the library), I would think that a good solution to this would be to pack the received response into the exception so that it could be inspected in the handling of the exception. Any thoughts?

**Comment by lovelydinosaur** [comment]
Interesting. Would you be able to share a full traceback that you get at the point the exception is raised?

I'd *assume* that from a code-flow perspective it probably occurs while the request body is still being sent in `send_body()`

https://github.com/encode/httpcore/blob/5809b95c45dc17843afebccf181c226c6ebe0c4c/httpcore/_async/http2.py#L348-L358

```python
    async def send_body(self, stream: AsyncByteStream, timeout: TimeoutDict) -> None:
        async for data in stream:
            while data:
                max_flow = await self.connection.wait_for_outgoing_flow(
                    self.stream_id, timeout
                )
                chunk_size = min(len(data), max_flow)
                chunk, data = data[:chunk_size], data[chunk_size:]
                await self.connection.send_data(self.stream_id, chunk, timeout)

        await self.connection.end_stream(self.stream_id, timeout)
```

At the point that it's waiting for more outgoing flow in order to complete the sending of the request, it'll be receiving incoming events...

https://github.com/encode/httpcore/blob/5809b95c45dc17843afebccf181c226c6ebe0c4c/httpcore/_async/http2.py#L168-L183

```python
    async def wait_for_outgoing_flow(self, stream_id: int, timeout: TimeoutDict) -> int:
        """
        Returns the maximum allowable outgoing flow for a given stream.
        If the allowable flow is zero, then waits on the network until
        WindowUpdated frames have increased the flow rate.
        https://tools.ietf.org/html/rfc7540#section-6.9
        """
        local_flow = self.h2_state.local_flow_control_window(stream_id)
        connection_flow = self.h2_state.max_outbound_frame_size
        flow = min(local_flow, connection_flow)
        while flow == 0:
            await self.receive_events(timeout)  # I'd assume this is the point at which we "see" the RST_STREAM
            local_flow = self.h2_state.local_flow_control_window(stream_id)
            connection_flow = self.h2_state.max_outbound_frame_size
            flow = min(local_flow, connection_flow)
        return flow
```

Anyways that's interesting from a POV of figuring out where in the codebase we'd want to tweak things if we wanted to handle this differently. But let's put that aside and move onto *what* we want the behaviour to be here.

Relevant part of [the HTTP/2 spec here](https://httpwg.org/specs/rfc7540.html#HttpSequence) is this...

> A server can send a complete response prior to the client sending an entire request if the response does not depend on any portion of the request that has not been sent and received. When this is true, a server MAY request that the client abort transmission of a request without error by sending a RST_STREAM with an error code of NO_ERROR after sending a complete response (i.e., a frame with the END_STREAM flag). Clients MUST NOT discard responses as a result of receiving such a RST_STREAM, though clients can always discard responses at their discretion for other reasons.

I think there's no reason for us to want to raise an exception in this case. The server didn't require the entire request body, and did send a response. Ideally we just want to return that response. It's immaterial that the entire request body wasn't ever fully consumed.

Perhaps we'll want to do something like this...

https://github.com/encode/httpcore/blob/5809b95c45dc17843afebccf181c226c6ebe0c4c/httpcore/_async/http2.py#L270-L300

```python
        await self.send_headers(method, url, headers, has_body, timeout)
        if has_body:
            try:
                await self.send_body(stream, timeout)
            except httpcore.RemoteProtocolError as exc:
                if exc.reset_stream and exc.error_code == ErrorCodes.NO_ERROR:
                    # If we get RST_STREAM with NO_ERROR then it indicates that
                    # the response has been sent without fully ingesting the request body.
                    # We can forget about sending the remainder of the request body
                    # here, and move on to parsing the response.
                    pass
                else:
                    raise exc

        # Receive the response.
        status_code, headers = await self.receive_response(timeout)
        response_stream = AsyncIteratorByteStream(
            aiterator=self.body_iter(timeout), aclose_func=self._response_closed
        )
```

**Reply by lovelydinosaur** [reply]
It also seems to me that we could easily get into a state where we *don't* ever see this error and just hang when sending the request...

The response data is also flow controlled, so if the response is large, then on the client side we might end up stuck waiting for body request flow control (which we never get, because the server has stopped ingesting the request body) and also hanging on response flow control (which we never get, because we on the client are not yet ingesting the response body), and so get to a point where we don't see the final DATA frame that includes the RST_STREAM/NO_ERROR.

Only ways around to deal with that is way more complex - restucturing the codebase so that we're performing parallel reading and writing of the request/response.

**Comment by chradcliffe** [comment]
This is the backtrace with httpx==0.17.1:

```
Traceback (most recent call last):
  File "<stripped path>/pypi__httpx/httpx/_exceptions.py", line 326, in map_exceptions
    yield
  File "<stripped path>/pypi__httpx/httpx/_client.py", line 1492, in _send_single_request
    (status_code, headers, stream, ext) = await transport.arequest(
  File "<stripped path>/pypi__httpx/httpx/_transports/default.py", line 169, in arequest
    return await self._pool.arequest(
  File "<stripped path>/pypi__httpcore/httpcore/_async/http_proxy.py", line 124, in arequest
    return await self._tunnel_request(
  File "<stripped path>/pypi__httpcore/httpcore/_async/http_proxy.py", line 258, in _tunnel_request
    (status_code, headers, stream, ext) = await connection.arequest(
  File "<stripped path>/pypi__httpcore/httpcore/_async/connection.py", line 106, in arequest
    return await self.connection.arequest(method, url, headers, stream, ext)
  File "<stripped path>/pypi__httpcore/httpcore/_async/http2.py", line 119, in arequest
    return await h2_stream.arequest(method, url, headers, stream, ext)
  File "<stripped path>/pypi__httpcore/httpcore/_async/http2.py", line 294, in arequest
    await self.send_body(stream, timeout)
  File "<stripped path>/pypi__httpcore/httpcore/_async/http2.py", line 355, in send_body
    max_flow = await self.connection.wait_for_outgoing_flow(
  File "<stripped path>/pypi__httpcore/httpcore/_async/http2.py", line 182, in wait_for_outgoing_flow
    await self.receive_events(timeout)
  File "<stripped path>/pypi__httpcore/httpcore/_async/http2.py", line 215, in receive_events
    raise RemoteProtocolError(event)
httpcore.RemoteProtocolError: <StreamReset stream_id:1, error_code:ErrorCodes.NO_ERROR, remote_reset:True>
```
